### PR TITLE
Update reverse infer to allow changing shape if it already partially defined

### PR DESCRIPTION
**Pull request by mvafin**
### Details:
 - *When shape is partially defined, e.g. rank is known, reverse infer wasn't updating shape even when it was possible. With this change it will do that.*

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: src/common/transformations/tests/common_optimizations/reverse_shape_and_type_infer.cpp [hunk 1]
```diff
@@ -43,6 +43,36 @@ TEST_F(TransformationTestsF, ConvolutionReverseInfer) {
...
+                                                           weights,
+                                                           Strides{2, 2},
+                                                           CoordinateDiff{3, 3},
+                                                           CoordinateDiff{3, 3},
+                                                           Strides{1, 1});
+        auto result = std::make_shared<opset10::Result>(conv);
+        model_ref = std::make_shared<Model>(ResultVector{result}, ParameterVector{data});
+    }
+}
+
```
Comment (slyalin) [inline src/common/transformations/tests/common_optimizations/reverse_shape_and_type_infer.cpp:75]: Could you also add a test with `Concat`? In my previous practice I needed it for concats.

BTW, the dimension deduction based on the lables should provide a more powerful way to do such a reverse propagation (actually the propagation in "any direction"). @jane-intel, what would you say?

Comment (mvafin) [reply src/common/transformations/tests/common_optimizations/reverse_shape_and_type_infer.cpp:75]: Test added

**Review by slyalin — APPROVED** [review]

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp [hunk 2]
```diff
@@ -247,28 +249,28 @@ bool ov::pass::ReverseShapeAndTypeInfer::run_on_model(const std::shared_ptr<ov::
...
             for (const auto& in_desc : else_in_desc) {
                 const auto& in_indx = in_desc->m_input_index;
                 const auto& body_indx = in_desc->m_body_parameter_index;
-                if (if_op->get_input_tensor(in_indx).get_partial_shape().rank().is_dynamic()) {
-                    if_op->get_input_tensor(in_indx).m_partial_shape =
-                        else_body_params.at(body_indx)->get_partial_shape();
+                auto else_p = else_body_params.at(body_indx);
+                if (if_op->get_input_tensor(in_indx).get_partial_shape().compatible(else_p->get_partial_shape())) {
+                    PartialShape::merge_into(if_op->get_input_tensor(in_indx).m_partial_shape,
+                                             else_p->get_partial_shape());
```
Comment (rkazants) [inline src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:269]: I am not sure that this is correct. For example, let us have `If` operation with `then-body` param of shape [-1] and `else-body` param shape `3` but it does not mean that the corresponding `if` input should have a shape [3]. It should have [-1] still

**Review by rkazants — COMMENTED** [review]
Need to be careful with these changes

**Review by rkazants — CHANGES_REQUESTED** [review]
need to discuss this offline due to comment above.

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp (hunk 2 above)
Comment (mvafin) [reply src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:269]: This code propagates shapes and types from body parameters to inputs of If operation, and those must have same shape and type for both bodies. The code for outputs isn't changed in this PR.

Comment (rkazants) [reply src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:269]: Discussed offline and agreed to re-consider changes in this PR

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp [hunk 3]
```diff
@@ -247,28 +249,28 @@ bool ov::pass::ReverseShapeAndTypeInfer::run_on_model(const std::shared_ptr<ov::
             for (const auto& in_desc : then_in_desc) {
                 const auto& in_indx = in_desc->m_input_index;
                 const auto& body_indx = in_desc->m_body_parameter_index;
-                if (if_op->get_input_tensor(in_indx).get_partial_shape().rank().is_dynamic()) {
-                    if_op->get_input_tensor(in_indx).m_partial_shape =
-                        then_body_params.at(body_indx)->get_partial_shape();
+                auto then_p = then_body_params.at(body_indx);
+                if (if_op->get_input_tensor(in_indx).get_partial_shape().compatible(then_p->get_partial_shape())) {
+                    PartialShape::merge_into(if_op->get_input_tensor(in_indx).m_partial_shape,
+                                             then_p->get_partial_shape());
```
Comment (mvafin) [inline src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:255]:
```suggestion
                if (if_op->get_input_tensor(in_indx).get_partial_shape().rank().is_dynamic()) {
                    if_op->get_input_tensor(in_indx).m_partial_shape =
                        then_body_params.at(body_indx)->get_partial_shape();
```

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp (hunk 2 above)
Comment (mvafin) [inline src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:269]:
```suggestion
                if (if_op->get_input_tensor(in_indx).get_partial_shape().rank().is_dynamic()) {
                    if_op->get_input_tensor(in_indx).m_partial_shape =
                        else_body_params.at(body_indx)->get_partial_shape();
```

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp [hunk 4]
```diff
@@ -247,28 +249,28 @@ bool ov::pass::ReverseShapeAndTypeInfer::run_on_model(const std::shared_ptr<ov::
...
+                auto else_p = else_body_params.at(body_indx);
+                if (if_op->get_input_tensor(in_indx).get_partial_shape().compatible(else_p->get_partial_shape())) {
+                    PartialShape::merge_into(if_op->get_input_tensor(in_indx).m_partial_shape,
+                                             else_p->get_partial_shape());
                     is_changed = true;
                 }
                 if (if_op->get_input_tensor(in_indx).get_element_type().is_dynamic()) {
-                    if_op->get_input_tensor(in_indx).m_element_type =
-                        else_body_params.at(body_indx)->get_element_type();
+                    if_op->get_input_tensor(in_indx).m_element_type = else_p->get_element_type();
```
Comment (mvafin) [inline src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:273]:
```suggestion
                    if_op->get_input_tensor(in_indx).m_element_type =
                        else_body_params.at(body_indx)->get_element_type();
```

**Review by rkazants — APPROVED** [review]

**Comment by rkazants** [comment]
I rebased branch due to recent regression in CPU

**Pull request merged**